pkg/kubelet/kuberuntime/kuberuntime_manager_test.go

```diff
@@ -657,6 +669,136 @@ func TestSyncPodWithInitContainers(t *testing.T) {
+		},
+	}
+	containers := []v1.Container{
+		{
+			Name:            "foo1",
+			Image:           "busybox",
+			ImagePullPolicy: v1.PullIfNotPresent,
+		},
+		{
+			Name:            "foo2",
+			Image:           "alpine",
+			ImagePullPolicy: v1.PullIfNotPresent,
+			Lifecycle: &v1.Lifecycle{
+				Type: &sidecarLifecyle,
+			},
+		},
+	}
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       "12345678",
+			Name:      "foo",
+			Namespace: "new",
+		},
+		Spec: v1.PodSpec{
+			Containers:     containers,
+			InitContainers: initContainers,
+		},
+	}
+
+	backOff := flowcontrol.NewBackOff(time.Second, time.Minute)
+
+	// 1. should only create the init container.
+	podStatus, err := m.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	assert.NoError(t, err)
+	result := m.SyncPod(pod, podStatus, []v1.Secret{}, backOff)
+	assert.NoError(t, result.Error())
+	expected := []*cRecord{
+		{name: initContainers[0].Name, attempt: 0, state: runtimeapi.ContainerState_CONTAINER_RUNNING},
+	}
+	verifyContainerStatuses(t, fakeRuntime, expected, "start only the init container")
+
+	// 2. should not create app container because init container is still running.
+	podStatus, err = m.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	assert.NoError(t, err)
+	result = m.SyncPod(pod, podStatus, []v1.Secret{}, backOff)
+	assert.NoError(t, result.Error())
+	verifyContainerStatuses(t, fakeRuntime, expected, "init container still running; do nothing")
+
+	// 3. should create all sidecar containers because init container finished.
+	// Stop init container instance 0.
+	sandboxIDs, err := m.getSandboxIDByPodUID(pod.UID, nil)
+	require.NoError(t, err)
+	sandboxID := sandboxIDs[0]
+	initID0, err := fakeRuntime.GetContainerID(sandboxID, initContainers[0].Name, 0)
+	require.NoError(t, err)
+	fakeRuntime.StopContainer(initID0, 0)
+	// Sync again.
+	pod.Status.ContainerStatuses = []v1.ContainerStatus{
+		{
+			Name:  "foo1",
+			State: v1.ContainerState{Waiting: &v1.ContainerStateWaiting{}},
+		},
+		{
+			Name:  "foo2",
+			State: v1.ContainerState{Waiting: &v1.ContainerStateWaiting{}},
+		},
+	}
+	podStatus, err = m.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	assert.NoError(t, err)
+	result = m.SyncPod(pod, podStatus, []v1.Secret{}, backOff)
+	assert.NoError(t, result.Error())
+	expected = []*cRecord{
+		{name: initContainers[0].Name, attempt: 0, state: runtimeapi.ContainerState_CONTAINER_EXITED},
+		{name: containers[1].Name, attempt: 0, state: runtimeapi.ContainerState_CONTAINER_RUNNING},
+	}
+	verifyContainerStatuses(t, fakeRuntime, expected, "init container completed; all sidecar containers should be running")
+
+	//4 Should start non-sidecars once sidecars are ready
+	pod.Status.ContainerStatuses = []v1.ContainerStatus{
+		{
+			Name:  "foo1",
+			Ready: true,
+			State: v1.ContainerState{Running: &v1.ContainerStateRunning{}},
+		},
+		{
+			Name:  "foo2",
+			State: v1.ContainerState{Waiting: &v1.ContainerStateWaiting{}},
```

Comment:
On line 685, container foo1 is declared as a normal container, and it is foo2 that is a sidecar. So, foo2 is ready before foo1, as it should. The test should be fixed in the following way:

Credits to @rata 

```
--- a/pkg/kubelet/kuberuntime/kuberuntime_manager_test.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_manager_test.go
@@ -760,11 +760,11 @@ func TestSyncPodWithSidecarsAndInitContainers(t *testing.T) {
        pod.Status.ContainerStatuses = []v1.ContainerStatus{
                {
                        Name:  "foo1",
-                       Ready: true,
                        State: v1.ContainerState{Running: &v1.ContainerStateRunning{}},
                },
                {
                        Name:  "foo2",
+                       Ready: true,
                        State: v1.ContainerState{Waiting: &v1.ContainerStateWaiting{}},
                },
        }
```

Reply:
yeah that's a mistake, good spot. However we also need to swap the states as well as readiness for this test, foo2 should be running and foo1 should be waiting. As we're meant to be testing that non-sidecars get started once the sidecars are running and ready.